File: clients/FSharp/Model/BuildProperties.fs
#nowarn "0058"
namespace Aicup2020.Model
type BuildProperties = {
    Options: EntityType[];
    InitHealth: option<int>;
    } with
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write this.Options.Length
        this.Options |> Array.iter (fun value ->
            writer.Write (int value)
        )
        match this.InitHealth with
            | Some value ->
                writer.Write true
                writer.Write value
            | None -> writer.Write false
    static member readFrom(reader: System.IO.BinaryReader) = {
        Options = [|for _ in 1 .. reader.ReadInt32() do
            yield reader.ReadInt32() |> enum
        |]
        InitHealth = match reader.ReadBoolean() with
            | true ->
                Some(
                    reader.ReadInt32()
                    )
            | false -> None
    }

File: clients/Scala/src/main/scala/model/BuildProperties.scala
package model

import util.StreamUtil

case class BuildProperties(options: Seq[model.EntityType], initHealth: Option[Int]) {
    def writeTo(stream: java.io.OutputStream) {
        StreamUtil.writeInt(stream, options.length)
        options.foreach { value =>
            value.writeTo(stream)
        }
        initHealth match {
            case None => StreamUtil.writeBoolean(stream, false)
            case Some(value) => {
                StreamUtil.writeBoolean(stream, true)
                StreamUtil.writeInt(stream, value)
            }
        }
    }
}
object BuildProperties {
    def readFrom(stream: java.io.InputStream): BuildProperties = BuildProperties(
        (0 until StreamUtil.readInt(stream)).map { _ =>
            model.EntityType.readFrom(stream)
        }
        ,
        if (StreamUtil.readBoolean(stream)) Some(
            StreamUtil.readInt(stream)
        ) else None
        )
}
